pkg/unsaferecovery/unsafe_recovery_controller.go

```diff
@@ -780,6 +781,12 @@ func (r *regionItem) IsRaftStale(origin *regionItem, u *Controller) bool {
 		func(a, b *regionItem) int {
 			return int(a.report.GetRaftState().GetHardState().GetTerm()) - int(b.report.GetRaftState().GetHardState().GetTerm())
 		},
+		// choose the peer has maximum applied_index or last index.
+		func(a, b *regionItem) int {
+			maxIdxA := typeutil.MaxUint64(a.report.GetRaftState().GetLastIndex(), a.report.AppliedIndex)
+			maxIdxB := typeutil.MaxUint64(b.report.GetRaftState().GetLastIndex(), b.report.AppliedIndex)
+			return int(maxIdxA - maxIdxB)
+		},
 		func(a, b *regionItem) int {
 			return int(a.report.GetRaftState().GetLastIndex()) - int(b.report.GetRaftState().GetLastIndex())
```

Comment:
Do we need to compare the last index again?

Reply:
Yes, if max(applied_index) == max(last_index), we better choose the max(last_index) peer to avoid doing force compact.